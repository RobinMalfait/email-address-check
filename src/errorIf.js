function errorIf(predicate, msg) {
  return async input => {
    // Resolve the error message
    if (typeof msg === "function") {
      msg = msg(input);
    }

    // Resolve the predicate
    if (typeof predicate === "function") {
      try {
        predicate = await predicate(input);
      } catch (err) {
        throw new Error(msg);
      }
    }

    // If the predicate is true, throw an error
    if (predicate) {
      throw new Error(msg);
    }
  };
}

export default errorIf;
