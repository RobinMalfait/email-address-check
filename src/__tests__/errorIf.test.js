import errorIf from "../errorIf";

describe("errorIf", () => {
  it("should throw an error, when the predicate condition is true", () => {
    const errorFn = errorIf(true, "the actual error message");

    return expect(errorFn("some input value")).rejects.toMatchSnapshot();
  });

  it("should throw an error, when the predicate condition is fn that resolves to true", () => {
    const errorFn = errorIf(() => true, "the actual error message");

    return expect(errorFn("some input value")).rejects.toMatchSnapshot();
  });

  it("should throw an error, when the predicate already throws an error", () => {
    const errorFn = errorIf(() => {
      throw new Error("An error in the condition");
    }, "the actual error message");

    return expect(errorFn("some input value")).rejects.toMatchSnapshot();
  });

  it("should throw an error, ", () => {
    const errorFn = errorIf(true, () => "the actual error message");

    return expect(errorFn("some input value")).rejects.toMatchSnapshot();
  });
});
