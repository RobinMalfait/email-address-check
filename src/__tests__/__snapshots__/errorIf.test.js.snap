// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorIf should throw an error,  1`] = `[Error: the actual error message]`;

exports[`errorIf should throw an error, when the predicate already throws an error 1`] = `[Error: the actual error message]`;

exports[`errorIf should throw an error, when the predicate condition is fn that resolves to true 1`] = `[Error: the actual error message]`;

exports[`errorIf should throw an error, when the predicate condition is true 1`] = `[Error: the actual error message]`;
